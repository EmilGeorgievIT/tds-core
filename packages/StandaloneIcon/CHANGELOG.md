# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.5...@tds/core-standalone-icon@1.1.0) (2019-06-04)

### Features

- **core-standalone-icon:** add innerRef prop to allow forwarding of ref ([fed0f6c](https://github.com/telusdigital/tds/commit/fed0f6c))

## [1.0.5](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.4...@tds/core-standalone-icon@1.0.5) (2019-05-01)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.4](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.3...@tds/core-standalone-icon@1.0.4) (2019-04-22)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.3](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.2...@tds/core-standalone-icon@1.0.3) (2019-03-18)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.1...@tds/core-standalone-icon@1.0.2) (2019-03-05)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.0...@tds/core-standalone-icon@1.0.1) (2019-02-06)

**Note:** Version bump only for package @tds/core-standalone-icon
