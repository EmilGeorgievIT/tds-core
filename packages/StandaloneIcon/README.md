# TDS Core: StandaloneIcon

An icon that has meaning within the context of the page, which should be communicated to screen readers.
