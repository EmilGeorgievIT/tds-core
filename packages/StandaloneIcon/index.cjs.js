require('./dist/index.css')
const StandaloneIcon = require('./dist/index.cjs')

module.exports = StandaloneIcon
