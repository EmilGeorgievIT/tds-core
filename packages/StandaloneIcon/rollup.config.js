import configure from '../../config/rollup.config'
import { dependencies } from './package.json'

export default configure({
  input: './StandaloneIcon.jsx',
  dependencies,
})
