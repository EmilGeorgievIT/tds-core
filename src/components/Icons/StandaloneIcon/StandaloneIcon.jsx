import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../utils/safeRest'

import Icon from '../Icon/Icon'

import styles from './StandaloneIcon.modules.scss'

/**
 * An icon that has meaning within the context of the page, which should be communicated to screen readers.
 *
 * <span class="docs--badge__new">new!</span> <span class="docs--badge__version">v0.24.0</span>
 */
class StandaloneIcon extends React.Component {
  onClick = event => {
    const {onClick} = this.props

    if (onClick) {
      onClick(event)
    }
  }

  render() {
    const {symbol, variant, size, interactive, onClick, a11yText, ...rest} = this.props

    if (interactive) {
      return (
        <button {...safeRest(rest)} className={styles.interactive} aria-label={a11yText}>
          <Icon symbol={symbol} variant={variant} size={size} />
        </button>
      )
    }

    return <Icon {...rest} symbol={symbol} variant={variant} size={size} aria-label={a11yText} />
  }
}

StandaloneIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
  /**
   * Make the icon interactive
   */
  interactive: PropTypes.bool,
  /**
  * A callback function to be invoked when the interactive icon is clicked.
  */
  onClick: PropTypes.func,
  /**
   * A description of the icon for screen readers.
   */
  a11yText: PropTypes.string.isRequired,
}

StandaloneIcon.defaultProps = {
  variant: undefined,
  size: 24,
  interactive: false,
  onClick: undefined,
}

export default StandaloneIcon
