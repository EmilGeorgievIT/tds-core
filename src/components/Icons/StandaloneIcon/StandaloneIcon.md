### Available Icons

```
<div className="docs--horizontal-spacing">
  <StandaloneIcon symbol="caretDown" a11yText="text" />
  <StandaloneIcon symbol="caretUp" a11yText="text" />
  <StandaloneIcon symbol="checkmark" a11yText="text" />
  <StandaloneIcon symbol="leftChevron" a11yText="text" />
  <StandaloneIcon symbol="chevron" a11yText="text" />
  <StandaloneIcon symbol="exclamationPointCircle" a11yText="text" />
  <StandaloneIcon symbol="expander" a11yText="text" />
  <StandaloneIcon symbol="hamburger" a11yText="text" />
  <StandaloneIcon symbol="incomplete" a11yText="text" />
  <StandaloneIcon symbol="location" a11yText="text" />
  <StandaloneIcon symbol="minus" a11yText="text" />
  <StandaloneIcon symbol="plus" a11yText="text" />
  <StandaloneIcon symbol="questionMarkCircle" a11yText="text" />
  <StandaloneIcon symbol="spyglass" a11yText="text" />
  <StandaloneIcon symbol="times" a11yText="text" />
</div>
```
