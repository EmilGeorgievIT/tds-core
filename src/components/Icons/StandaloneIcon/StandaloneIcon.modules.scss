.interactive {
  appearance: none;
  background: none;
  border: 0;
  cursor: pointer;
}
