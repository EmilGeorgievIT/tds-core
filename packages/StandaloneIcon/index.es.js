import './dist/index.css'
import StandaloneIcon from './dist/index.es'

export default StandaloneIcon
