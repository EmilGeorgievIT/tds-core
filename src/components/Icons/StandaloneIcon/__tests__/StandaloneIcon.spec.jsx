import React from 'react'
import { shallow, render } from 'enzyme'


import StandaloneIcon from '../StandaloneIcon'

describe('StandaloneIcon', () => {
  describe('Plain StandaloneIcon', () => {
    const defaultProps = {
      symbol: 'spyglass',
      a11yText: 'Some text for the screen readers.',
    }
    const doShallow = (props = {}) => shallow(<StandaloneIcon {...defaultProps} {...props} />)

    it('renders', () => {
      const icon = render(<StandaloneIcon {...defaultProps} />)
      expect(icon).toMatchSnapshot()
    })

    it('passes attributes to the Icon component', () => {
      const icon = doShallow({
        symbol: 'checkmark',
        variant: 'secondary',
        size: 16,
        id: 'the-icon',
      })

      expect(icon).toHaveProp('symbol', 'checkmark')
      expect(icon).toHaveProp('variant', 'secondary')
      expect(icon).toHaveProp('size', 16)
      expect(icon).toHaveProp('id', 'the-icon')
    })

    it('is accessible for screen readers', () => {
      const icon = doShallow({ a11yText: 'Some screen reader text' })

      expect(icon).not.toHaveProp('aria-hidden')
      expect(icon).toHaveProp('aria-label', 'Some screen reader text')
    })

    it('does not allow custom CSS', () => {
      const icon = doShallow({
        className: 'my-custom-class',
        style: { color: 'hotpink' },
      })

      expect(icon).not.toHaveProp('className', 'my-custom-class')
      expect(icon).not.toHaveProp('style')
    })
  })

  describe('Interactive StandaloneIcon', () => {
    const defaultProps = {
      symbol: 'spyglass',
      a11yText: 'Some text for the screen readers.',
      onClick: jest.fn(),
    }
    const doShallow = (props = {}) => shallow(<StandaloneIcon {...defaultProps} {...props} />)

    it('renders', () => {
      const interactiveIcon = render(<StandaloneIcon {...defaultProps} />)

      expect(interactiveIcon).toMatchSnapshot()
    })

    it('renders an HTML button tag', () => {
      const interactiveIcon = doShallow({ onClick: jest.fn() })

      expect(interactiveIcon).toHaveTagName('button')
    })

    it('triggers the click handler', () => {
      const onClickMock = jest.fn()
      const interactiveIcon = doShallow({ onClick: onClickMock })

      interactiveIcon.simulate('click')

      expect(onClickMock).toHaveBeenCalled()
    })

    describe('touch area', () => {
      it('expands the touch area to be large enough for a finger when the icon is small', () => {
        let interactiveIcon = doShallow({ size: 16 })
        expect(interactiveIcon).toHaveStyle('padding', '8px')
        expect(interactiveIcon).toHaveStyle('margin', '-8px')

        interactiveIcon = doShallow({ size: 24 })

        expect(interactiveIcon).toHaveStyle('padding', '4px')
        expect(interactiveIcon).toHaveStyle('margin', '-4px')
      })

      it('does not expand the touch area when the icon is large', () => {
        const interactiveIcon = doShallow({ size: 48 })

        expect(interactiveIcon).not.toHaveStyle('padding')
        expect(interactiveIcon).not.toHaveStyle('margin')
      })
    })

    it('passes additional attributes to the button element', () => {
      const interactiveIcon = doShallow({
        id: 'the-interactiveIcon',
        role: 'button',
      })

      expect(interactiveIcon).toHaveProp('id', 'the-interactiveIcon')
      expect(interactiveIcon).toHaveProp('role', 'button')
    })

    it('does not allow custom CSS', () => {
      const interactiveIcon = doShallow({
        className: 'my-custom-class',
        style: { color: 'hotpink' },
      })

      expect(interactiveIcon).not.toHaveProp('className', 'my-custom-class')
      expect(interactiveIcon).not.toHaveStyle('color', 'hotpink')
    })
  })
})
