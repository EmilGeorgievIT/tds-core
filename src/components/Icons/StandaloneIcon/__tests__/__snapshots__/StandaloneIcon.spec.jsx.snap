// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandaloneIcon renders 1`] = `
<i
  aria-label="Some screen reader text."
  class="iconSpyglass size24"
>
  
</i>
`;
